app/frontend/src/features/messages/groupchats/AdminsDialog.tsx

```diff
@@ -0,0 +1,249 @@
+import {
+  CircularProgress,
+  DialogProps,
+  List,
+  ListItem,
+} from "@material-ui/core";
+import { Empty } from "google-protobuf/google/protobuf/empty_pb";
+import { Error as GrpcError } from "grpc-web";
+import React, { useEffect, useState } from "react";
+import { useMutation, useQueryClient } from "react-query";
+
+import Alert from "../../../components/Alert";
+import Avatar from "../../../components/Avatar";
+import Button from "../../../components/Button";
+import ConfirmationDialogWrapper from "../../../components/ConfirmationDialogWrapper";
+import {
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogTitle,
+} from "../../../components/Dialog";
+import IconButton from "../../../components/IconButton";
+import { AddIcon, CloseIcon } from "../../../components/Icons";
+import TextBody from "../../../components/TextBody";
+import { User } from "../../../pb/api_pb";
+import { GroupChat } from "../../../pb/conversations_pb";
+import { service } from "../../../service";
+import { useAuthContext } from "../../auth/AuthProvider";
+import useUsers from "../../userQueries/useUsers";
+import { useMembersDialogStyles } from "./MembersDialog";
+
+function AdminListItem({
+  groupChatId,
+  member,
+  memberIsAdmin,
+  setError,
+}: {
+  groupChatId: number;
+  member: User.AsObject;
+  memberIsAdmin: boolean;
+  setError: (value: string) => void;
+}) {
+  const classes = useMembersDialogStyles();
+
+  const isCurrentUser = useAuthContext().authState.userId === member.userId;
+
+  const queryClient = useQueryClient();
+  const clearError = () => setError("");
+  const handleError = (error: GrpcError) => setError(error.message);
+  const invalidate = () => {
+    queryClient.invalidateQueries(["groupChatMessages", groupChatId]);
+    queryClient.invalidateQueries(["groupChats"]);
+    queryClient.invalidateQueries(["groupChat", groupChatId]);
+  };
+
+  const makeAdmin = useMutation<Empty, GrpcError, void>(
+    () => service.conversations.makeGroupChatAdmin(groupChatId, member),
+    {
+      onMutate: clearError,
+      onSuccess: () => {
+        const previousGroupChat = queryClient.getQueryData<GroupChat.AsObject>([
+          "groupChat",
+          groupChatId,
+        ]);
+        const newAdminUserIdsList = previousGroupChat?.adminUserIdsList ?? [];
+        newAdminUserIdsList.push(member.userId);
+        queryClient.setQueryData(["groupChat", groupChatId], {
+          ...previousGroupChat,
+          adminUserIdsList: newAdminUserIdsList,
+        });
```

Comment:
Optimistic update. Otherwise it stops loading, but the user stays there until the invalidation is completed.

Reply:
It won't be a problem because it only sets the data on success. It's more a "semi-optimistic" update. 

I could do it their way though, and it would look instant to the user. What do you think?

Reply:
Is the delay noticeable if you didn't go with this "semi-optimistic" route?

Reply:
Even not a slow connection, because the loading spinner stops (because the mutation is complete), but there a short pause while the query refetches on invalidation, then the user jumps to the correct place.

Reply:
Seems like a worthy tradeoff for better UX but without the faff of having to worry about rolling back overly optimistic updates, let's stick with it 👍